core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1343,21 +1346,31 @@ public boolean maybeAcquireFetchLock() {
     }
 
     /**
-     * Release the fetch lock once the records are fetched from the leader.
+     * Release the fetch lock once the records are fetched from the leader. It is imperative that the DelayedShareFetch instance
+     * that acquired the fetch lock should be the one releasing it.
+     * @param fetchId - The DelayedShareFetch instance uuid that is trying to release the fetch lock.
      */
-    void releaseFetchLock() {
+    void releaseFetchLock(Uuid fetchId) {
         // Register the metric for the duration the fetch lock was held. Do not register the metric
         // if the fetch lock was not acquired.
-        if (fetchLock.get()) {
-            long currentTime = time.hiResClockMs();
+        long currentTime = time.hiResClockMs();
+        if (fetchLock.compareAndSet(fetchId, null)) {
             long acquiredDurationMs = currentTime - fetchLockAcquiredTimeMs;
             // Update the metric for the fetch lock time.
             sharePartitionMetrics.recordFetchLockTimeMs(acquiredDurationMs);
             // Update fetch lock ratio metric.
             recordFetchLockRatioMetric(acquiredDurationMs);
             fetchLockReleasedTimeMs = currentTime;
+        } else {
+            // This code should not be reached unless we are in error-prone scenarios. Since we are releasing the fetch
+            // lock for multiple share partitions at different places in DelayedShareFetch (due to tackling remote
+            // storage fetch and local log fetch from a single purgatory), in order to safeguard ourselves from bad code,
+            // we are logging when an instance that does not hold the fetch lock tries to release it.
+            log.info("Instance {} does not hold the fetch lock, yet trying to release it for share partition {}-{}",
+                fetchId, groupId, topicIdPartition);
+            fetchLock.set(null);
```

Comment:
If you used `AtomicReference.getAndSet()` here, you would receive the identity of the current holder. That might be interesting for debugging.